mixer/pkg/attribute/envoyProtoBag.go

```diff
@@ -0,0 +1,316 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	"net"
+	"strings"
+	"fmt"
+	"sync"
+	"sort"
+	authzGRPC "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	accessLogGRPC "github.com/envoyproxy/go-control-plane/envoy/service/accesslog/v2"
+	"time"
+	mixerpb "istio.io/api/mixer/v1"
+	attr "istio.io/pkg/attribute"
+
+)
+
+// Exports of types for backwards compatibility
+//type (
+//	Bag        = attr.Bag
+//	MutableBag = attr.MutableBag
+//)
+
+// ProtoBag implements the Bag interface on top of an Attributes proto.
+type EnvoyProtoBag struct {
+	reqMap              map[string]interface{}
+	//proto               *mixerpb.CompressedAttributes
+	globalDict          map[string]int32
+	globalWordList      []string
+	messageDict         map[string]int32
+	convertedStringMaps map[int32]attr.StringMap
+	stringMapMutex      sync.RWMutex
+
+	// to keep track of attributes that are referenced
+	referencedAttrs      map[attr.Reference]attr.Presence
+	referencedAttrsMutex sync.Mutex
+}
+
+// referencedAttrsSize is the size of referenced attributes.
+
+var envoyProtoBags = sync.Pool{
+	New: func() interface{} {
+		return &EnvoyProtoBag{
+			referencedAttrs: make(map[attr.Reference]attr.Presence, referencedAttrsSize),
+		}
+	},
+}
+
+// GetProtoBag returns a proto-based attribute bag.
+// When you are done using the proto bag, call the Done method to recycle it.
+func GetEnvoyProtoBagAuthz(req *authzGRPC.CheckRequest) *EnvoyProtoBag {
+	//, globalDict map[string]int32,	globalWordList []string) *EnvoyProtoBag {
+	pb := envoyProtoBags.Get().(*EnvoyProtoBag)
+
+	// build the message-level dictionary
+	reqMap := make(map[string]interface{})
+	//TODO account for other protocols
+	reqMap["context.protocol"] = "http"
+	//req_map["context.proxy_version"] = "unknown"
+	reqMap["context.reporter.kind"] = "server"
+	//req_map["context.reporter.uid"] = "unknown"
+	reqMap["destination.ip"] = []byte(net.ParseIP(req.GetAttributes().GetDestination().GetAddress().GetSocketAddress().
```

Comment:
i know that `AttributeContext` and `HttpLogs` are vastly different protos, but there is a lot overlap here.

Are there common areas that we can extract out the logic to a single place?  Like maybe `PopulateAddressAttributes(prefix string, address core.Address, attributes map[string]string) map[string]string` ?  And maybe something similar for Http attributes?

Reply:
Between authz and ALS the only place I could see condensing working is after GetSocketAddress() they all get Address and Port the same way so that could condense it a bit. They also have the same ip address conversions so I could condense it a bit like that. Beyond condensing those two in a function I'm not sure if there would be any other one that it would be effective for.

I could condense ALS httpLogs and tcpLogs code significantly though since they share 6 or so attributes almost identically.